pkg/controller/pod_volume_restore_controller.go

```diff
@@ -160,64 +101,106 @@ func (c *podVolumeRestoreController) pvrHandler(obj interface{}) {
-	pod := obj.(*corev1api.Pod)
-	log := c.logger.WithField("key", kube.NamespaceAndName(pod))
+	patchHelper, err := patch.NewHelper(pvr, c.Client)
+	if err != nil {
+		log.WithError(err).Error("Unable to new patch helper")
+		return ctrl.Result{}, err
+	}
 
-	// the pod should always be for this node since the podInformer is filtered
-	// based on node, so this is just a failsafe.
-	if !isPodOnNode(pod, c.nodeName) {
-		return
+	log.Info("Restore starting")
+	pvr.Status.Phase = velerov1api.PodVolumeRestorePhaseInProgress
+	pvr.Status.StartTimestamp = &metav1.Time{Time: c.clock.Now()}
+	if err = patchHelper.Patch(ctx, pvr); err != nil {
+		log.WithError(err).Error("Unable to update status to in progress")
+		return ctrl.Result{}, err
 	}
+	if err = c.processRestore(ctx, pvr, pod, log); err != nil {
+		pvr.Status.Phase = velerov1api.PodVolumeRestorePhaseFailed
+		pvr.Status.Message = err.Error()
+		pvr.Status.CompletionTimestamp = &metav1.Time{Time: c.clock.Now()}
+		if e := patchHelper.Patch(ctx, pvr); e != nil {
+			log.WithError(err).Error("Unable to update status to failed")
+		}
 
-	if !isResticInitContainerRunning(pod) {
-		log.Debug("Pod is not running restic-wait init container, not enqueuing restores for pod")
-		return
+		log.WithError(err).Error("Unable to process the PodVolumeRestore")
+		return ctrl.Result{}, err
 	}
 
-	resticInitContainerIndex := getResticInitContainerIndex(pod)
-	if resticInitContainerIndex > 0 {
-		log.Warnf(`Init containers before the %s container may cause issues
-		          if they interfere with volumes being restored: %s index %d`, restic.InitContainer, restic.InitContainer, resticInitContainerIndex)
+	pvr.Status.Phase = velerov1api.PodVolumeRestorePhaseCompleted
+	pvr.Status.CompletionTimestamp = &metav1.Time{Time: c.clock.Now()}
+	if err = patchHelper.Patch(ctx, pvr); err != nil {
+		log.WithError(err).Error("Unable to update status to completed")
+		return ctrl.Result{}, err
 	}
+	log.Info("Restore completed")
+	return ctrl.Result{}, nil
+}
 
-	selector := labels.Set(map[string]string{
-		velerov1api.PodUIDLabel: string(pod.UID),
-	}).AsSelector()
+func (c *PodVolumeRestoreReconciler) shouldProcess(ctx context.Context, log logrus.FieldLogger, pvr *velerov1api.PodVolumeRestore) (bool, *corev1api.Pod, error) {
+	if !isPVRNew(pvr) {
+		log.Debug("PodVolumeRestore is not new, skip")
+		return false, nil, nil
+	}
 
-	pvrs, err := c.podVolumeRestoreLister.List(selector)
-	if err != nil {
-		log.WithError(err).Error("Unable to list pod volume restores")
-		return
+	// we filter the pods during the initialization of cache, if we can get a pod here, the pod must be in the same node with the controller
+	// so we don't need to compare the node anymore
+	pod := &corev1api.Pod{}
+	if err := c.Get(ctx, types.NamespacedName{Namespace: pvr.Spec.Pod.Namespace, Name: pvr.Spec.Pod.Name}, pod); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.WithError(err).Debug("Pod not found on this node, skip")
```

Comment:
I think `on this node` is a little misleading, b/c at this point the pod is not found via api-server?

Reply:
The client gets pod from cache and the cache contains only the pods in this node, so if we get not found error, it just means the pod isn't on this node, but it may exist on other nodes

Reply:
make sense.